GitUI/CommandsDialogs/BrowseDialog/FormOpenDirectory.cs

```diff
@@ -16,16 +16,11 @@ public FormOpenDirectory(GitModule? currentModule)
             InitializeComponent();
             InitializeComplete();
 
-            ThreadHelper.JoinableTaskFactory.Run(async () =>
-            {
-                var repositoryHistory = await RepositoryHistoryManager.Locals.LoadRecentHistoryAsync();
-
-                await this.SwitchToMainThreadAsync();
```

Comment:
> General: How can you ensure that the UI thread is used in all these occurrences?

By manual review: Form / Page ctors, UI event handlers / callbacks are to called from the UI thread.
I can add `ThreadHelper.ThrowIfNotOnUIThread()` in locations where we are not fully sure.

Reply:
> ThreadHelper.ThrowIfNotOnUIThread()

Yes, that would be good.
Refactoring may try to use these snippets, so better to add too many than too few.

Reply:
I think it may be a good idea to add those assertions now to catch possible regressions. And after awhile, we can remove those.

Reply:
> I think it may be a good idea to add those assertions now to catch possible regressions. And after awhile, we can remove those.

In the short run, assertions are not needed - code review must handle that calls are equivalent.

Comments and assertions are needed for longer term maintenance.

Reply:
> > I think it may be a good idea to add those assertions now to catch possible regressions. And after awhile, we can remove those.
> 
> In the short run, assertions are not needed - code review must handle that calls are equivalent.
> 
> Comments and assertions are needed for longer term maintenance.

Most of my reviews (for better or worse) are done in a browser, sadly I often don't have the luxury of checking the code flow in a IDE. So, for me an added assertion is a way to boost a confidence level.

Reply:
Aren't forms always required to be constructed on the UI thread?

Reply:
> Aren't forms always required to be constructed on the UI thread?

yes, see above comments

> This assertion, when present in a synchronous method, should generally be the first call in the method

I have added the `ThreadHelper.AssertOnUIThread()` on explicit request - at the same location as `SwitchToMainThreadAsync()`. I would rather remove them again than move them to the beginning.

Reply:
> I would rather remove them again than move them to the beginning.

Standard practice for vs-threading is to place prerequisite threading assertions early in the method. The general rule is synchronous methods may assert at the beginning, while asynchronous methods should switch if necessary. The pair of analyzer rules related to this are:

https://github.com/microsoft/vs-threading/blob/fcfe71f8c8e05c84f8185e9773164e107c4e9666/doc/analyzers/VSTHRD108.md
https://github.com/microsoft/vs-threading/blob/fcfe71f8c8e05c84f8185e9773164e107c4e9666/doc/analyzers/VSTHRD109.md

Reply:
The assertitions should be done where needed, normally not by the caller.
This is fine here,  see comment on GitUI/CommandsDialogs/BrowseDialog/FormRecentReposSettings.cs.

I prefer that assertion to be called where needed in the routine, but if there is a standard here, it should be used here too.